src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java

```diff
@@ -111,221 +91,178 @@ public List<Route> routes() {
+                    .map(username -> loadConfigurationWithRequestContent(username, request))
+                    .map(endpointValidator::isAllowedToChangeImmutableEntity)
+                    .map(this::validateSecurityRoles)
+                    .map(securityConfiguration -> createOrUpdateAccount(request, securityConfiguration))
+                    .map(this::validateAndUpdatePassword)
+                    .map(this::addEntityToConfig)
+            );
+    }
 
-        final ObjectNode contentAsNode = (ObjectNode) content;
-        final SecurityJsonNode securityJsonNode = new SecurityJsonNode(contentAsNode);
+    ValidationResult<String> withAuthTokenPath(final RestRequest request) throws IOException {
+        return endpointValidator.withRequiredEntityName(nameParam(request)).map(username -> {
+            // Handle auth token fetching
+            if (!(request.uri().contains("/internalusers/" + username + "/authtoken") && request.uri().endsWith("/authtoken"))) {
+                return ValidationResult.error(RestStatus.NOT_IMPLEMENTED, methodNotImplementedMessage(request.method()));
+            }
+            return ValidationResult.success(username);
+        });
+    }
 
-        // Don't allow user to add non-existent role or a role for which role-mapping is hidden or reserved
-        final List<String> securityRoles = securityJsonNode.get("opendistro_security_roles").asList();
-        if (securityRoles != null) {
-            for (final String role : securityRoles) {
-                if (!isValidRolesMapping(channel, role)) {
-                    return;
-                }
+    void generateAuthToken(final RestChannel channel, final SecurityConfiguration securityConfiguration) throws IOException {
+        try {
+            final var username = securityConfiguration.entityName();
+            final var authToken = userService.generateAuthToken(username);
+            if (!Strings.isNullOrEmpty(authToken)) {
+                ok(channel, "'" + username + "' authtoken generated " + authToken);
+            } else {
+                badRequest(channel, "'" + username + "' authtoken failed to be created.");
             }
+        } catch (final UserServiceException e) {
+            badRequest(channel, e.getMessage());
         }
+    }
 
-        final boolean userExisted = internalUsersConfiguration.exists(username);
-
-        // when updating an existing user password hash can be blank, which means no
-        // changes
+    ValidationResult<SecurityConfiguration> validateSecurityRoles(final SecurityConfiguration securityConfiguration) throws IOException {
+        return loadConfiguration(CType.ROLES, false, false).map(rolesConfiguration -> {
+            final var contentAsNode = (ObjectNode) securityConfiguration.requestContent();
+            final var securityJsonNode = new SecurityJsonNode(contentAsNode);
+            // FIXME do we need to verify roles as well?
```

Comment:
Looks liked there is this and a couple other FIXME's in this PR.  What do you want to do with these create issue to follow up or address before merging?

Reply:
@peternied Thank you,
I'm planning to address them after PR is merged as soon as possible. Regarding this FIXME i will remove it. I misunderstood this part, we have `security_roles`, `backend_roles`, etc., so we do not need validate other roles except `security_roles`.

Reply:
Having one place to store all future efforts as issues in github is my preference, but ain't no big deal for small items.